### ISSUE TITLE
Is there a way to add cover art when there's an error or it's not found?

### ISSUE THREAD

**Issue by Scotty-Trees** [Taiko2k/TauonMusicBox#387]
Greetings again! Absolutely loving this flatpak version! Having said that, I do have some thoughts! (Austin Powers reference, sorry I couldn't help myself!)

[ ISSUE ]
I had a few blank/black missing album artworks that Tauon didn't pick up. But I managed to figure out if I changed my view style to showcase+lyrics I could use a right click menu option that could quick-fetch cover art. This worked and added a few missing artworks I had which was great. This right click menu option is a bit hidden though, I think it needs to be added to the other view types as well if possible. I did however come across two issues when doing this though. 

The first issue was trying to add a Crown The Empire - Sudden Sky artwork from the quick fetch menu, which actually crashes the app every single time. It found the artwork was showing in the showcase+lyrics view only, but that wasn't the correct artwork and in the Gallery view it just shows a blank/black cover art and since that's the view I mainly use I assumed the artwork was missing. So I attempted to use the showcase+lyrics view to delete the artwork and try to quick add the artwork again cuz why not sometimes things like that work, but it simply crashes when I do this. Only case I've had in this regard, every other add artwork worked fine. Terminal output of that is below. 

```
scott@pop-os:~$ flatpak run com.github.taiko2k.tauonmb
Tauon Music Box v6.3.2
Copyright 2015-2020 Taiko2k captain.gxj@gmail.com

Gtk-Message: 14:11:03.337: Failed to load module "appmenu-gtk-module"
Detected running as Flatpak
Installed as Snap
User files location: /home/scott/.var/app/com.github.taiko2k.tauonmb/data/TauonMusicBox
Found XDG-Downloads: /home/scott/Downloads
Found XDG-Music: /home/scott/Music
Install directory: /app/bin
Database loaded in 0.081 seconds.
Applying scale based on xft setting
Using UI scale: 1.0
Could not connect to gnome media keys
Init GStreamer...
Using SDL version: 2.0.12
Applying theme: Carbon
ERROR: Image load failed on track: /home/scott/Music/OnePlus Music/Distinguisher - Hell from Here (2019)/01. Actuality.mp3
ERROR: Image load failed on track: 
- /home/scott/Music/OnePlus Music/Distinguisher - Hell from Here (2019)/01. Actuality.mp3
Auto save playtime
ERROR: Image load failed on track: /home/scott/Music/OnePlus Music/Crown The Empire - Sudden Sky (2019)/01. (X).mp3
ERROR: Image load failed on track: 
- /home/scott/Music/OnePlus Music/Crown The Empire - Sudden Sky (2019)/01. (X).mp3
ERROR: Image load failed on track: /home/scott/Music/OnePlus Music/Crown The Empire - Sudden Sky (2019)/01. (X).mp3
ERROR: Image load failed on track: 
- /home/scott/Music/OnePlus Music/Crown The Empire - Sudden Sky (2019)/01. (X).mp3
image file is truncated (3 bytes not processed)
Thumbnail error
/home/scott/Music/OnePlus Music/Crown The Empire - Sudden Sky (2019)/01. (X).mp3
Attempting to connect to Last.fm network
Connection to lastfm appears successful
Using tagged release group ID: 7b4330e1-a799-444c-9f4b-71623c10b6ac
Using tagged artist ID: 5cf3bae7-166f-44c2-b48f-63ddea9d4fb2
Deleted file: /home/scott/Music/OnePlus Music/Crown The Empire - Sudden Sky (2019)/cover-307613.jpg
image file is truncated (3 bytes not processed)
Thumbnail error
/home/scott/Music/OnePlus Music/Crown The Empire - Sudden Sky (2019)/03. what i am.mp3
Using tagged release group ID: 7b4330e1-a799-444c-9f4b-71623c10b6ac
Using tagged artist ID: 5cf3bae7-166f-44c2-b48f-63ddea9d4fb2
Traceback (most recent call last):
  File "/app/bin/tauon.py", line 10397, in display
    im = im.resize(new_size, Image.ANTIALIAS)
  File "/app/lib/python3.7/site-packages/PIL/Image.py", line 1890, in resize
    self.load()
  File "/app/lib/python3.7/site-packages/PIL/ImageFile.py", line 249, in load
    "(%d bytes not processed)" % len(b)
OSError: image file is truncated (3 bytes not processed)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/bin/tauon.py", line 39483, in <module>
    showcase.render()
  File "/app/bin/tauon.py", line 34949, in render
    album_art_gen.display(track, (x, y), (box, box))
  File "/app/bin/tauon.py", line 10615, in display
    del self.source_cache[index][offset]
KeyError: 4929
```
The second issue I came across is when I tried adding artwork for Distinguisher - Hell From Here, which the quick-fetcher was unable to find. But I do have a cover.jpg of it in the albums folder, which sometimes shows up in the showcase+lyrics view, but it's always blank/black in Gallery view. I've tried re-adding, re-scanning my music library, but it still shows up blank for this. I know it's probably not much to go on, but here's the terminal output for this as well.

```
scott@pop-os:~$ flatpak run com.github.taiko2k.tauonmb
Tauon Music Box v6.3.2
Copyright 2015-2020 Taiko2k captain.gxj@gmail.com

Gtk-Message: 14:20:53.874: Failed to load module "appmenu-gtk-module"
Detected running as Flatpak
Installed as Snap
User files location: /home/scott/.var/app/com.github.taiko2k.tauonmb/data/TauonMusicBox
Found XDG-Downloads: /home/scott/Downloads
Found XDG-Music: /home/scott/Music
Install directory: /app/bin
Database loaded in 0.083 seconds.
Applying scale based on xft setting
Using UI scale: 1.0
Could not connect to gnome media keys
Init GStreamer...
Using SDL version: 2.0.12
image file is truncated (2 bytes not processed)
Thumbnail error
/home/scott/Music/OnePlus Music/Distinguisher - Hell from Here (2019)/01. Actuality.mp3
Applying theme: Carbon
ERROR: Image load failed on track: /home/scott/Music/OnePlus Music/Distinguisher - Hell from Here (2019)/01. Actuality.mp3
ERROR: Image load failed on track: 
- /home/scott/Music/OnePlus Music/Distinguisher - Hell from Here (2019)/01. Actuality.mp3
Auto save playtime
Attempting to connect to Last.fm network
Connection to lastfm appears successful
MusicBrainz ID lookup...
Found release group ID: acef9622-1e5a-3f40-9111-a54441c24634
Found artist ID: d2d67fcb-df35-41ee-8ee5-a347e0e8bb33
Failed to get from fanart.tv
Writing database to disk.
Unloading SDL...
SDL unloaded
GStreamer unloaded
bye
```

[ Final Thoughts ]

Truth be told, these are in no way deal breaker issues at all. These are very minor issues. These issues will not prevent me from using this wonderful music app. I've been clutching onto Rhythmbox for like a decade and you've managed to get me to switch, so congrats! If you can tackle these issues great, but if not I fully understand. 

If there's anything to take away from this issue, I would like to have the following questions answered or clarified so I understand this lovely music app just a little bit more.

1.) Is it possible to add the "quick-fetch cover art" right click option to the "Gallery" view since that's my personal favorite view style by far to use with this application. It would be great to right click an artwork in Gallery view and quick fetch artwork that is missing. Right now it appears it only shows up in the showcase+lyrics view. I do really love how the artwork loads up instantly when scrolling btw, great work. 

2.) How does Tauon read artwork? Does it read the embedded image within an audio track or does it just go by whatever it finds via musicbrainz, last.fm, etc.? Not all, but most of my music folders have a cover.jpg within the folder of the album. Is Tauon able to read the album cover artwork that way, or no?

3. Is there a way or will there be a way to have the ability to find/add our own artwork from a local folder? Sorry if Tauon already does this and I'm just a fool!

I think that's all I have for now, hope this helps in any way and I look forward to your reply and thanks very much again for making what is now one of my favorite Linux applications!

**Comment by Taiko2k** [comment]
> 1) Is it possible to add the "quick-fetch cover art" right click option to the "Gallery" view

The intended main way of accessing this menu is the `Tracks + Art` view. (The top one in the view selector). Its kind of intended that your album art management is done there and not in the gallery.

> 2) How does Tauon read artwork?

It reads both embedded artwork and loose image files in the same folder (or in an appropriately named subfolder). In both `Tracks + Art` and `Showcase + Lyrics`, you can click the artwork to cycle between all the found images. The one selected there will be the one used for the gallery.

It only ever attempts to find art on the internet when you press `Quick fetch cover art`.

> 3) Is there a way or will there be a way to have the ability to find/add our own artwork from a local folder?

Yep, it does this as mentioned above, just add the image to the same folder as the track.

> Issue

It probably shouldn't crash like that so that can be regarded as a bug, though it seems to have something to do with the downloaded image being corrupted. Perhaps that particular one just isn't going to work.

As a side note, new art you add yourself might not show up in the gallery immediately. Restarting the app should fix this. (Or you can go `MENU` > `Database` > `Reset Image Cache`)

**Comment by Scotty-Trees** [comment]
Ok, it took a bit of trial and error, but I finally figured out a little fix that works. I'm not sure why, but out of hundreds of albums, Tauon didn't display 2 of them properly in Gallery view, but oddly enough they did show up in the other views like showcase+lyrics and tracks+art. It took a bit of tinkering between adding different cover art files to the folders that were missing it/not showing it, but I've come to a fix of sorts at least for my problem.

So when on the 1% chance that Tauon doesn't display the album art work correctly between the various views, I would make sure I had a cover art file in the albums music folder. And also, I would fetch the cover artwork from Tauon, so that the album would now have 2 cover artworks. Since one of them doesn't want to show for whatever reason, just like you said you can choose different artworks via the Tracks+Art view, so I did that and it worked. That's only part one of a fix though. That only works when Tauon does fetch the album cover and it's also in my music folder. What happens when Tauon (ie. musicbrainz) can't find the album artwork, what then? Well since I had that issue happen (only once on hundreds of albums, so it's a very small error!), my fix was to have two covers in the albums music folder, cover.jpg and cover1.jpg. Tauon recognizes there's two in the folder, so even though the first one doesn't want to show in all views, by choosing the second cover, it displayed the artwork correctly in all the views. 
Totally not sure why this is happening, but like I said it's a very minor issue and it seems to very rarely occur. Manually having to edit a few album artworks is no big deal. But now that I now know how to fix it, I'd say this issue is good for me, other than that one bug where it crashes sometimes when you try to fetch or delete an album artwork. I couldn't replicate it exactly, sometimes it would and sometimes it wouldn't but I don't think it's a bug that happens often. If I notice it more or anything I'll be sure to let you know. 
Keep up the great work man, I really can't say enough how much using your music app is just genuinely fun to use, and with the gallery view i'm discovering a bit of old music I've missed, since I normally listen to music that was recently added. Cheers for now man, but good luck in tackling the weird little crash bug, if you need any testing for a new version, I'll be available for sure!

**Comment by Taiko2k** [comment]
Does that issue with the art not showing in the gallery persist after a restart of the app?

**Comment by Scotty-Trees** [comment]
IF I don't utilize the "fix" I mentioned above, then yes the issue will persist even after the app is restarted, and even after the image cache is reset numerous times. 

BUT if I use my "fix" (I hesitant to call it that, more like a workaround really) and I have 2 album covers in the folder, Tauon will approve of the second one and no matter how many times I restart, the album art will show as intended in all views. 

Now, a bit of a kicker is, whenever I set the image cache to reset, this issue will have to be re-fixed on my end. It's only 3 albums that randomly do this, but it's always the same 3 albums. Hope that tid bit helps!

**Comment by Taiko2k** [comment]
Okay I think I have a better idea of what might be going on now.

Its likely that your music files have an embedded picture, but the data inside is corrupted. The standard album art display falls back to the next picture, but the gallery still tries and fails to load it.

So, the proper workaround would be to try clean your music files of those albums in a tag editor, removing or replacing the embedded picture. If you do try that let me know how that goes.

**Comment by Scotty-Trees** [comment]
Okay looks like that solves it! I might have forgot to mention, I'm currently using Pop!_OS, but not all gtk tag editors work that well so I used Kid3 (qt) to remove the embedded artwork in the files of those 3 albums that seemed to cause the issue. All 3 albums were cleaned, and I left a cover.jpg in each album folder, so Tauon picked up those first. To verify it all worked alright, I restarted the app and reset the image cache a couple times, and luckily the correct album art was showing across all view types. No more blank/black cover artwork anywhere to be found! Such a simple solution, I'm surprised I hadn't already attempted that myself, but hindsight right! 

Thank you very much for taking the time to help me with this issue.
